File: backend/app/schemas/drill.py
"""
Schemas Pydantic para Drill, Play y Tag.
"""
from __future__ import annotations

from datetime import datetime
from typing import Any

from pydantic import BaseModel, Field

from app.models.drill import CourtLayoutType, DrillType


# ── Tag ───────────────────────────────────────────────────────────────────────

class TagCreate(BaseModel):
    name: str = Field(..., min_length=1, max_length=100)
    color: str | None = Field(None, pattern=r"^#[0-9a-fA-F]{6}$")


class TagUpdate(BaseModel):
    name: str | None = Field(None, min_length=1, max_length=100)
    color: str | None = Field(None, pattern=r"^#[0-9a-fA-F]{6}$")


class TagResponse(BaseModel):
    id: int
    user_id: int
    name: str
    color: str | None
    archived_at: datetime | None
    created_at: datetime

    model_config = {"from_attributes": True}


# ── Drill / Play ───────────────────────────────────────────────────────────────

class DrillCreate(BaseModel):
    type: DrillType
    name: str = Field(..., min_length=1, max_length=255)
    court_layout: CourtLayoutType = CourtLayoutType.half_fiba
    description: str | None = None
    # root_sequence se inicializa en el backend con un nodo raíz vacío
    tag_ids: list[int] = Field(default_factory=list)


class DrillUpdate(BaseModel):
    name: str | None = Field(None, min_length=1, max_length=255)
    court_layout: CourtLayoutType | None = None
    description: str | None = None
    root_sequence: dict[str, Any] | None = None
    tag_ids: list[int] | None = None  # None = no tocar tags; [] = quitar todos


class DrillResponse(BaseModel):
    id: int
    user_id: int
    type: DrillType
    name: str
    court_layout: CourtLayoutType
    description: str | None
    root_sequence: dict[str, Any]
    parent_id: int | None
    archived_at: datetime | None
    created_at: datetime
    updated_at: datetime
    tags: list[TagResponse]

    model_config = {"from_attributes": True}


class DrillSummaryResponse(BaseModel):
    """Versión compacta sin root_sequence (para listados)."""
    id: int
    user_id: int
    type: DrillType
    name: str
    court_layout: CourtLayoutType
    description: str | None
    parent_id: int | None
    archived_at: datetime | None
    created_at: datetime
    updated_at: datetime
    tags: list[TagResponse]

    model_config = {"from_attributes": True}

File: shared/types/drill.ts
// ── Enums ─────────────────────────────────────────────────────────────────────

export type DrillType = "drill" | "play";

export type CourtLayoutType =
  | "full_fiba"
  | "half_fiba"
  | "mini_fiba"
  | "half_mini_fiba";

export const COURT_LAYOUT_LABELS: Record<CourtLayoutType, string> = {
  full_fiba: "Cancha completa FIBA",
  half_fiba: "Media cancha FIBA",
  mini_fiba: "Cancha mini FIBA (3×3)",
  half_mini_fiba: "Media cancha mini FIBA",
};

// ── Sketch elements ───────────────────────────────────────────────────────────

export type ElementType =
  | "player_offense"
  | "player_defense"
  | "ball"
  | "cone"
  | "basket"
  | "line_move"
  | "line_dribble"
  | "line_pass"
  | "line_screen"
  | "line_cut"
  | "text";

export interface Point {
  x: number;
  y: number;
}

export interface SketchElement {
  id: string;
  type: ElementType;
  /** Posición normalizada [0.0, 1.0] */
  x: number;
  y: number;
  rotation: number;
  /** ID estable del jugador al que pertenece este elemento (RF-185) */
  playerId?: string;
  /** Para líneas de movimiento: jugador que ejecuta el movimiento (RF-187) */
  ownerId?: string;
  /** Puntos de la línea (para tipos line_*) */
  points?: Point[];
  color?: string;
  symbol?: string;
  label?: string;
  lineStyle?: "solid" | "dashed" | "dotted";
  arrowHead?: boolean;
}

// ── Sequence tree ─────────────────────────────────────────────────────────────

export interface SequenceNode {
  id: string;
  elements: SketchElement[];
  branches: SequenceNode[];
  label: string | null;
}

// ── Tag ───────────────────────────────────────────────────────────────────────

export interface Tag {
  id: number;
  user_id: number;
  name: string;
  color: string | null;
  archived_at: string | null;
  created_at: string;
}

export interface TagCreate {
  name: string;
  color?: string | null;
}

export interface TagUpdate {
  name?: string;
  color?: string | null;
}

// ── Drill / Play ──────────────────────────────────────────────────────────────

export interface DrillSummary {
  id: number;
  user_id: number;
  type: DrillType;
  name: string;
  court_layout: CourtLayoutType;
  description: string | null;
  parent_id: number | null;
  is_favorite: boolean;
  archived_at: string | null;
  created_at: string;
  updated_at: string;
  tags: Tag[];
  variant_count: number;
}

export interface Drill extends DrillSummary {
  root_sequence: SequenceNode;
}

export interface DrillCreate {
  type: DrillType;
  name: string;
  court_layout?: CourtLayoutType;
  description?: string | null;
  tag_ids?: number[];
}

export interface DrillUpdate {
  name?: string;
  court_layout?: CourtLayoutType;
  description?: string | null;
  root_sequence?: SequenceNode;
  tag_ids?: number[];
}
